File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.js
const fs = require('fs');
// 백준 제출용
const input = fs.readFileSync(0).toString().trim().split('\n');

// 로컬 테스트용
// const filePath =
//   process.platform === 'linux' ? 'dev/stdin' : __dirname + '/input.txt';
// let input = fs.readFileSync(filePath).toString().trim().split('\n');

let n = Number(input[0].split(' ')[0]);
let m = Number(input[0].split(' ')[1]);
let data = input[1].split(' ').map(Number);

let end = 0;
let count = 0;
let sum = 0;

/*
 * two-pointer 진행
 */
for (let start = 0; start < n; start++) {
  // 현재 부분의 합이 target 보다 작다면 end를 1 증가
  while (sum < m && end < n) {
    sum += data[end];
    end++;
  }

  if (sum == m) count++;

  // 현재 부분의 합이 target 보다 크다면 start를 1 증가
  sum -= data[start];
}

console.log(count);

File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		int[] data = new int[N];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			data[i] = Integer.parseInt(st.nextToken());
		}

		// 시작점과 끝점이 첫번째 원소의 인덱스를 가리키도록 함
		int end = 0, count = 0;
		int sum = 0;

		// start 증가시키면서 포인터 이동
		for (int start = 0; start < N; start++) {
			// 현재 부분의 합이 target 보다 작다면 end를 1 증가
			while (sum < M && end < N) {
				sum += data[end];
				end++;
			}

			if (sum == M) count++;

			// 현재 부분의 합이 target 보다 크다면 start를 1 증가
			sum -= data[start]; // start가 증가하기 전에 sum에서 start 제거 필요
		}


		System.out.println(count);
	}

//	static String input = "5 5\n"
//			+ "1 2 3 4 5";
}